File: src/main/java/me/sargunvohra/leveluphp/command/SetIntCommand.java
package me.sargunvohra.leveluphp.command;

import static net.minecraft.command.Commands.literal;

import com.mojang.brigadier.arguments.IntegerArgumentType;
import com.mojang.brigadier.builder.LiteralArgumentBuilder;
import java.util.function.BiConsumer;
import lombok.experimental.UtilityClass;
import lombok.val;
import me.sargunvohra.leveluphp.capability.PlayerLevelHandler;
import net.minecraft.command.CommandSource;
import net.minecraft.command.Commands;
import net.minecraft.command.arguments.EntityArgument;

@UtilityClass
class SetIntCommand {
  public LiteralArgumentBuilder<CommandSource> register(
      String literal, int permissionLevel, BiConsumer<PlayerLevelHandler, Integer> executor) {
    val command = literal(literal).requires(source -> source.hasPermissionLevel(permissionLevel));

    command.then(
        Commands.argument("amount", IntegerArgumentType.integer(0))
            .executes(
                context -> {
                  val player = context.getSource().asPlayer();
                  val amount = IntegerArgumentType.getInteger(context, "amount");

                  player
                      .getCapability(PlayerLevelHandler.CAPABILITY)
                      .ifPresent(handler -> executor.accept(handler, amount));

                  return 0;
                }));

    command.then(
        Commands.argument("players", EntityArgument.players())
            .then(
                Commands.argument("amount", IntegerArgumentType.integer(0))
                    .executes(
                        context -> {
                          val players = EntityArgument.getPlayers(context, "players");
                          val amount = IntegerArgumentType.getInteger(context, "amount");

                          for (val player : players) {
                            player
                                .getCapability(PlayerLevelHandler.CAPABILITY)
                                .ifPresent(handler -> executor.accept(handler, amount));
                          }

                          return 0;
                        })));

    return command;
  }
}

File: src/main/java/me/sargunvohra/leveluphp/command/SetIntCommand.kt
package me.sargunvohra.leveluphp.command

import com.mojang.brigadier.arguments.IntegerArgumentType.getInteger
import com.mojang.brigadier.arguments.IntegerArgumentType.integer
import com.mojang.brigadier.builder.LiteralArgumentBuilder
import me.sargunvohra.leveluphp.core.PlayerLevelHandler
import me.sargunvohra.leveluphp.core.playerLevelHandler
import net.minecraft.command.CommandSource
import net.minecraft.command.Commands.argument
import net.minecraft.command.Commands.literal
import net.minecraft.command.arguments.EntityArgument.getPlayers
import net.minecraft.command.arguments.EntityArgument.players

internal class SetIntCommand {
    companion object {

        fun register(
            literal: String,
            permissionLevel: Int,
            set: (PlayerLevelHandler, Int) -> Unit
        ): LiteralArgumentBuilder<CommandSource> {
            return literal(literal)
                .requires { it.hasPermissionLevel(permissionLevel) }
                .then(
                    argument("amount", integer(0))
                        .executes { ctx ->
                            val player = ctx.source.asPlayer()
                            val amount = getInteger(ctx, "amount")

                            player.playerLevelHandler.ifPresent {
                                set(it, amount)
                            }

                            return@executes 0
                        }
                )
                .then(argument("players", players())
                    .then(argument("amount", integer(0))
                        .executes { ctx ->
                            val players = getPlayers(ctx, "players")
                            val amount = getInteger(ctx, "amount")

                            players.forEach { player ->
                                player.playerLevelHandler.ifPresent {
                                    set(it, amount)
                                }
                            }

                            return@executes 0
                        }
                    )
                )
        }
    }
}
